### PR TITLE
✨ Add email notification for transfer & purchase

### PR DESCRIPTION
**Pull request by nwingt**
Todos: 
- [x] Add setting in setting page
- [x] Confirm setting name in firestore
- [x] Refine email template
- [x] Add unsubscribe link

### Notification settings
image

### Transfer notification email
image

### Purchase notification email
image

### CODE REVIEW

File: src/locales/zh-Hant.json [hunk 1]
```diff
@@ -469,6 +469,15 @@
   "settings_following_login_in_button": "驗證",
   "settings_following_login_label": "驗證錢包以查看追蹤列表",
   "settings_liker_id": "Liker ID",
+  "settings_notification": "通知",
+  "settings_notification_fetch_error_message": "載入通知設定時出現錯誤：{error}",
+  "settings_notification_title": "通知設定",
+  "settings_notification_tx_purchase": "收藏",
+  "settings_notification_tx_purchase_price_threshold": "當價格大於上述設定值時便會發出通知",
```
Comment (williamchong) [inline src/locales/zh-Hant.json:476]:
```suggestion
  "settings_notification_tx_purchase_price_threshold": "當交易價格大於上述設定值時便會發出通知",

```

File: src/components/CheckBox.vue [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      ]"
+    >
+      <input
+        :class="[
+          'absolute',
+          'w-0',
+          'h-0',
+          'opacity-0',
+          'inset',
+        ]"
```
Comment (williamchong) [inline src/components/CheckBox.vue:17]: Passing `disabled` might be useful?

File: src/server/api/middleware/auth.js [hunk 3]
```diff
@@ -7,6 +9,33 @@ function authenticateV2Login(req, res, next) {
...
+    res.status(400).send('ADDRESS_MISMATCH');
+    return;
+  }
+  next();
+}
+
+async function checkEmailHasVerified(req, res, next) {
+  const { wallet } = req.params;
+  const userDoc = await walletUserCollection.doc(wallet).get();
+  const { email, emailUnconfirmed } = userDoc.data();
```
Comment (williamchong) [inline src/server/api/middleware/auth.js:25]: Do we need to guard `userDoc.data()` not exists?

File: src/server/api/routes/users/v2/notification.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  }
+});
+
+router.post(
+  '/notification',
+  authenticateV2Login,
+  checkEmailHasVerified,
+  async (req, res, next) => {
+    try {
+      setPrivateCacheHeader(res);
```
Comment (williamchong) [inline src/server/api/routes/users/v2/notification.js:35]: We should set private cache header in `authenticateV2Login`? Both security and code styling sense it seems to be an improvement
Also setting private cache for POST seems redundant

File: src/server/api/routes/users/v2/notification.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+const { walletUserCollection } = require('../../../../modules/firebase');
+
+const router = Router();
+
+router.get('/notification', authenticateV2Login, async (req, res, next) => {
+  try {
+    const { user } = req.session;
+    const userDoc = await walletUserCollection.doc(user).get();
+    const { notification = {} } = userDoc.data();
+    const { transfer = true, purchasePrice = 0 } = notification;
```
Comment (williamchong) [inline src/server/api/routes/users/v2/notification.js:17]: Should we call `setPrivateCacheHeader` here instead?

File: src/components/CheckBox.vue (hunk 2 above)
Comment (williamchong) [reply src/components/CheckBox.vue:17]: Seems `isDisabled` is used but not implemented

File: functions/pubsub/handlers.js [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+  const { name, image } = await fetchNFTMetadata(classId);
+
+  const userDoc = await walletUserCollection.doc(toAddress).get();
+  const userDocData = userDoc.data();
+  if (!userDocData) return;
+
+  const {
+    email,
+    notification: { transfer = false } = {},
```
Comment (williamchong) [inline functions/pubsub/handlers.js:227]: transfer is default to `true` in GET api, but `false` here, potential inconsistency? Maybe should be `true` here also?

**Review by williamchong — CHANGES_REQUESTED** [review]

File: src/server/api/middleware/auth.js (hunk 3 above)
Comment (WeiJun0827) [reply src/server/api/middleware/auth.js:25]: Logically we should have created a doc for a logged-in user, else the request would be blocked by the `authenticateV2Login` middleware before calling `checkEmailHasVerified`?

Comment (WeiJun0827) [reply src/server/api/middleware/auth.js:25]: But that's the case if we always check `authenticateV2Login` first... 🌚

**Review by williamchong — APPROVED** [review]

**Review by williamchong — APPROVED** [review]

**Pull request merged**